### PR TITLE
Make aframe work with WebVR APIs 1.0

### PR DESCRIPTION
**Pull request by dmarcos**
I made the necessary changes to adapt aframe to the new API spec.

- I added two flags to the webvr-polyfill to disable the UI and turn-your-phone instructions. The package.json points to my fork while I submit and merge a PR on boris repo.
- I tested on mobile (Safari iOS / Chrome for Android) and desktop Windows on Firefox (no webvr) and latest WebVR enabled Chromium builds on Rift and Vive. Everything seems to work.

### CODE REVIEW

File: src/style/aframe.css [hunk 1]
```diff
@@ -6,6 +6,10 @@ html {
   top: 0;
 }
 
+:-webkit-full-screen {
```
Comment (dmarcos) [inline src/style/aframe.css]: in chrome mobile the user agent stylesheet set it to white.

File: src/core/scene/fullscreen.js [hunk 2]
```diff
@@ -1,64 +0,0 @@
...
-}
-
-function iframedFullscreenChangeHandler (event) {
-  var scene = this;
-  if (!event.data) { return; }
-
-  switch (event.data.type) {
-    case 'fullscreen': {
-      switch (event.data.data) {
-        case 'enter':
```
Comment (cvan) [inline src/core/scene/fullscreen.js]: I think we should port this logic: some postMessage handler for entering and exiting using new API

File: src/core/scene/fullscreen.js [hunk 3]
```diff
@@ -1,64 +0,0 @@
...
-          exitFullscreenHandler(scene);
-          break;
-      }
-    }
-  }
-}
-
-function enterFullscreenHandler (scene) {
-  scene.addState('fullscreen');
-  scene.emit('fullscreen-enter');
```
Comment (cvan) [inline src/core/scene/fullscreen.js]: did you update the docs?

File: src/index.js [hunk 4]
```diff
@@ -1,6 +1,21 @@
 // Polyfill `Promise`.
 window.Promise = window.Promise || require('promise-polyfill');
 
+// Check before the polyfill runs
+window.hasNativeWebVRImplementation = !!navigator.getVRDisplays;
```
Comment (cvan) [inline src/index.js]: I would change Implementation to Support

File: src/index.js [hunk 5]
```diff
@@ -1,6 +1,21 @@
...
+
+// WebVR polyfill configuration.
+window.hasNonPolyfillWebVRSupport = !!navigator.getVRDevices || !!navigator.getVRDisplays;
+window.WebVRConfig = window.WebVRConfig || {
+  CARDBOARD_UI_DISABLED: true,
```
Comment (cvan) [inline src/index.js]: sucky, eh

File: src/index.js [hunk 6]
```diff
@@ -1,6 +1,21 @@
 // Polyfill `Promise`.
 window.Promise = window.Promise || require('promise-polyfill');
 
+// Check before the polyfill runs
+window.hasNativeWebVRImplementation = !!navigator.getVRDisplays;
+
+// WebVR polyfill configuration.
+window.hasNonPolyfillWebVRSupport = !!navigator.getVRDevices || !!navigator.getVRDisplays;
```
Comment (cvan) [inline src/index.js]: IMO we should rename to hasNativeLegacyWebVRSupport

File: src/core/scene/fullscreen.js [hunk 7]
```diff
@@ -1,64 +0,0 @@
...
-function fullscreenChangeHandler (event) {
-  var fullscreenElement = document.fullscreenElement ||
-                          document.mozFullScreenElement ||
-                          document.webkitFullscreenElement;
-  var scene = this;
-
-  // Lock to landscape orientation on mobile.
-  if (scene.isMobile && window.screen.orientation) {
-    if (fullscreenElement) {
-      window.screen.orientation.lock('landscape');
```
Comment (cvan) [inline src/core/scene/fullscreen.js]: can you confirm that this is still correctly handled by the polyfill?

File: src/core/scene/a-scene.js [hunk 8]
```diff
@@ -72,13 +79,11 @@ var AScene = module.exports = registerElement('a-scene', {
 
     attachedCallback: {
       value: function () {
-        initFullscreen(this);
         initMetaTags(this);
         initWakelock(this);
 
         window.addEventListener('load', this.resize.bind(this));
         window.addEventListener('resize', this.resize.bind(this), false);
-        this.addEventListener('fullscreen-exit', this.exitVR.bind(this));
```
Comment (cvan) [inline src/core/scene/a-scene.js]: per my previous PR, we ought to exit VR upon `beforeunload`

Comment (cvan) [inline src/core/scene/a-scene.js]: also, I hadn't gotten far with it - but I think we need to have an Exit VR button

File: package.json [hunk 9]
```diff
@@ -36,7 +36,7 @@
     "style-attr": "^1.0.2",
     "three": "^0.75.0",
     "tween.js": "^15.0.0",
-    "webvr-polyfill": "borismus/webvr-polyfill#3f47796"
+    "webvr-polyfill": "dmarcos/webvr-polyfill#70de389"
```
Comment (cvan) [inline package.json]: are you going to upstream?

File: package.json [hunk 10]
```diff
@@ -36,7 +36,7 @@
     "style-attr": "^1.0.2",
     "three": "^0.75.0",
```
Comment (cvan) [inline package.json]: you need to update to dev's VREffect. Lots of fixes lately

**Comment by cvan** [comment]
we should add keyboard shortcuts per https://github.com/cvan/aframe/commit/af2cfa39fe8bc72cc1875009f810f7e3c8560f13#diff-b4f74a9a2c02669787c5210f0cbed271R20

File: src/core/scene/a-scene.js [hunk 11]
```diff
@@ -128,20 +133,15 @@ var AScene = module.exports = registerElement('a-scene', {
...
       value: function (event) {
-        this.setStereoRenderer();
-        if (isMobile) {
-          setFullscreen(this.canvas);
-        } else {
-          this.stereoRenderer.setFullScreen(true);
-        }
+        if (!vrDisplay) { return; }
+        vrDisplay.requestPresent([{source: this.renderer.domElement}]);
         this.addState('vr-mode');
```
Comment (cvan) [inline src/core/scene/a-scene.js]: we should not blindly assume that this succeeded. we should change state and emit events only when the promise resolves à la https://github.com/cvan/aframe/commit/af2cfa39fe8bc72cc1875009f810f7e3c8560f13#diff-eb9bb1084863a2f65678ea67d6e7cf5eR135

File: src/core/scene/a-scene.js [hunk 12]
```diff
@@ -128,20 +133,15 @@ var AScene = module.exports = registerElement('a-scene', {
...
         this.emit('enter-vr', event);
       }
     },
 
     exitVR: {
       value: function () {
-        this.setMonoRenderer();
         this.removeState('vr-mode');
```
Comment (cvan) [inline src/core/scene/a-scene.js]: we should be calling `exitPresent`, right?

and we also need to wait for promise to resolve: https://github.com/cvan/aframe/commit/af2cfa39fe8bc72cc1875009f810f7e3c8560f13#diff-eb9bb1084863a2f65678ea67d6e7cf5eR147

**Comment by cvan** [comment]
thanks for doing this. I really appreciate it. I don't have my CV1 in front of me atm but I'll test as soon as I get a chance.

**Comment by cvan** [comment]
and just make sure we are using the latest VREffect and VRControls. also see this: https://github.com/mrdoob/three.js/issues/8568#issuecomment-215463445 it's important

File: src/index.js (hunk 6 above)
Comment (dmarcos) [inline src/index.js]: Actually I think we don't need this variable anymore.

File: package.json (hunk 9 above)
Comment (dmarcos) [inline package.json]: Submitted https://github.com/borismus/webvr-polyfill/pull/97

File: src/components/look-controls.js [hunk 13]
```diff
@@ -109,75 +117,76 @@ module.exports.Component = registerComponent('look-controls', {
...
-      return hmdQuaternion;
+      var el = this.el;
+      var currentPosition = el.getComputedAttribute('position');
+      var currentHMDPosition;
+      var previousHMDPosition = this.previousHMDPosition;
+      if (!this.data.hmdEnabled) { return; }
+      currentHMDPosition = this.calculateHMDPosition();
+      deltaHMDPosition.copy(currentHMDPosition).sub(previousHMDPosition);
+      previousHMDPosition.copy(currentHMDPosition);
+      // Do nothing if we have not moved
```
Comment (cvan) [inline src/components/look-controls.js]: punctuation`.`

File: src/core/scene/a-scene.js [hunk 14]
```diff
@@ -72,13 +80,12 @@ var AScene = module.exports = registerElement('a-scene', {
...
     attachedCallback: {
       value: function () {
-        initFullscreen(this);
         initMetaTags(this);
         initWakelock(this);
 
         window.addEventListener('load', this.resize.bind(this));
+        window.addEventListener('beforeunload', this.exitVR.bind(this));
         window.addEventListener('resize', this.resize.bind(this), false);
-        this.addEventListener('fullscreen-exit', this.exitVR.bind(this));
```
Comment (cvan) [inline src/core/scene/a-scene.js]: why are we removing this event listener?

File: src/core/scene/a-scene.js [hunk 15]
```diff
@@ -72,13 +80,12 @@ var AScene = module.exports = registerElement('a-scene', {
 
     attachedCallback: {
       value: function () {
-        initFullscreen(this);
         initMetaTags(this);
         initWakelock(this);
 
         window.addEventListener('load', this.resize.bind(this));
```
Comment (cvan) [inline src/core/scene/a-scene.js]: why don't we set an intermediate variable (e.g., `var resize = this.resize.bind(this);`)?

File: src/core/scene/a-scene.js [hunk 16]
```diff
@@ -72,13 +80,12 @@ var AScene = module.exports = registerElement('a-scene', {
...
+        window.addEventListener('beforeunload', this.exitVR.bind(this));
         window.addEventListener('resize', this.resize.bind(this), false);
```
Comment (cvan) [inline src/core/scene/a-scene.js]: you can remove the third argument - `false`. it's the default.

File: src/core/scene/a-scene.js [hunk 17]
```diff
@@ -128,22 +135,32 @@ var AScene = module.exports = registerElement('a-scene', {
...
       value: function (event) {
-        this.setStereoRenderer();
-        if (isMobile) {
-          setFullscreen(this.canvas);
-        } else {
-          this.stereoRenderer.setFullScreen(true);
+        var self = this;
+        var vrSource = [{source: this.renderer.domElement}];
+        if (!vrDisplay) { return; }
+        vrDisplay.requestPresent(vrSource).then(enterVRSuccess, enterVRFailure);
```
Comment (cvan) [inline src/core/scene/a-scene.js]: can you return the `Promise`?

File: src/core/scene/a-scene.js [hunk 18]
```diff
@@ -128,22 +135,32 @@ var AScene = module.exports = registerElement('a-scene', {
...
+        function enterVRSuccess () {
+          self.addState('vr-mode');
+          self.emit('enter-vr', event);
+        }
+        function enterVRFailure () {
+          error('enter VR mode error. requestPresent failed');
```
Comment (cvan) [inline src/core/scene/a-scene.js]: you don't want to print the error?

File: src/core/scene/a-scene.js [hunk 19]
```diff
@@ -128,22 +135,32 @@ var AScene = module.exports = registerElement('a-scene', {
...
       value: function () {
-        this.setMonoRenderer();
-        this.removeState('vr-mode');
-        this.emit('exit-vr', { target: this });
+        var self = this;
+        if (!vrDisplay) { return; }
+        vrDisplay.exitPresent().then(exitVRSuccess, exitVRFailure);
+        function exitVRSuccess () {
+          self.removeState('vr-mode');
+          self.emit('exit-vr', { target: self });
```
Comment (cvan) [inline src/core/scene/a-scene.js]: spaces around `{` and `}` necessary?

File: tests/__init.test.js [hunk 20]
```diff
@@ -5,6 +5,24 @@
  */
 window.debug = true;
 
+// getVRDisplays mock
+navigator.getVRDisplays = function () {
```
Comment (cvan) [inline tests/__init.test.js]: this better?

```js
navigator.getVRDisplays = function () {
  this.requestPresent = Promise.resolve;
  this.exitPresent = Promise.resolve;
};
```

Comment (dmarcos) [inline tests/__init.test.js]: Can you expand on this? How is the code you propose equivalent? If I want to mock ```getVRDisplays``` I need to return a promise right? Your function returns undefined.

Comment (cvan) [inline tests/__init.test.js]: sorry:

```js
navigator.getVRDisplays = function () {
  this.requestPresent = Promise.resolve;
  this.exitPresent = Promise.resolve;
  return Promise.resolve([{}, {}]);  // Resolve an array of `VRDisplay`s.
};
```

Comment (dmarcos) [inline tests/__init.test.js]: This stil does not make it. You are no attaching requestPreset/exitPresent to a mock vrDisplay that the promise will return.

File: src/core/scene/a-scene.js (hunk 14 above)
Comment (dmarcos) [inline src/core/scene/a-scene.js]: On WebVR 1.0. We don't rely on the fullscreen API anymore to enter/exit VR

File: src/index.js (hunk 4 above)
Comment (dmarcos) [inline src/index.js]: Implementation conveys better the meaning than support. Browsers ship or not a native implementation of the APIs. Support has an ambiguous meaning in this context.

File: src/core/scene/a-scene.js (hunk 18 above)
Comment (dmarcos) [inline src/core/scene/a-scene.js]: What do you mean? The error will be shown in console. Do you mean show on screen?

File: tests/__init.test.js (hunk 20 above)
Comment (cvan) [inline tests/__init.test.js]: I don't understand what you are saying

File: tests/components/scene/vr-mode-ui.test.js [hunk 21]
```diff
@@ -9,8 +9,16 @@ suite('vr-mode-ui', function () {
   setup(function (done) {
     this.entityEl = entityFactory();
     var el = this.el = this.entityEl.parentNode;
+    var resolvePromise = function () {
+      return new Promise(function (resolve) {
```
Comment (ngokevin) [inline tests/components/scene/vr-mode-ui.test.js]: `Promise.resolve()` will return a resolved promise

File: src/style/aframe.css (hunk 1 above)
Comment (ngokevin) [inline src/style/aframe.css]: Worth a comment

File: src/components/scene/canvas.js [hunk 22]
```diff
@@ -26,8 +26,8 @@ module.exports.Component = register('canvas', {
     if (!canvas) {
       canvas = document.createElement('canvas');
       canvas.classList.add('a-canvas');
-      canvas.style.height = data.height + '%';
-      canvas.style.width = data.width + '%';
+      canvas.style.height = data.height + 'px';
```
Comment (ngokevin) [inline src/components/scene/canvas.js]: `%` vs `px`?

File: src/core/scene/a-scene.js [hunk 23]
```diff
@@ -10,7 +9,9 @@ var utils = require('../../utils/');
 // Require after.
 var AEntity = require('../a-entity');
 var ANode = require('../a-node');
+var initPostMessageAPI = require('./postMessageAPI');
 
+var error = utils.debug('scene:error');
```
Comment (ngokevin) [inline src/core/scene/a-scene.js]: If it's an error, it always make sense to throw an actual error

File: src/core/scene/a-scene.js [hunk 24]
```diff
@@ -128,22 +129,42 @@ var AScene = module.exports = registerElement('a-scene', {
      */
     enterVR: {
       value: function (event) {
-        this.setStereoRenderer();
-        if (isMobile) {
-          setFullscreen(this.canvas);
-        } else {
-          this.stereoRenderer.setFullScreen(true);
+        var self = this;
+        var promise = this.effect.requestPresent();
```
Comment (ngokevin) [inline src/core/scene/a-scene.js]: More specific variable name? This is sort of like `var int = 5;`

File: examples/test/iframe/index.html [hunk 25]
```diff
@@ -0,0 +1,58 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title>Cube</title>
```
Comment (ngokevin) [inline examples/test/iframe/index.html]: cube

File: docs/core/scene.md [hunk 26]
```diff
@@ -41,17 +41,17 @@ The scene inherits from the [`Entity`][entity] class so it inherits all of its p
 
 | Name    | Description                                                                                                            |
 |---------|------------------------------------------------------------------------------------------------------------------------|
-| enterVR | Switch to stereo renderer and enter fullscreen. Needs to be called within a user-generated event handler like `click`. |
-| exitVR  | Switch to mono renderer and exit fullscreen.                                                                           |
+| enterVR | Switch to stereo render and pushes content to the headset. Needs to be called within a user-generated event handler like `click`. the first time a page enters VR. |
```
Comment (ngokevin) [inline docs/core/scene.md]: s/pushes/push to keep consistent tense.

there's an extra period after click

File: docs/core/scene.md [hunk 27]
```diff
@@ -41,17 +41,17 @@ The scene inherits from the [`Entity`][entity] class so it inherits all of its p
...
+| exitVR  | Switch to mono renderer and stops presenting content on the headset.                                                                           |
```
Comment (ngokevin) [inline docs/core/scene.md]: s/stops/stop

File: docs/core/scene.md [hunk 28]
```diff
@@ -41,17 +41,17 @@ The scene inherits from the [`Entity`][entity] class so it inherits all of its p
...
 | reload  | Revert the scene to its original state.                                                                                |
 
 ## Events
 
 | Name         | Description                         |
 |--------------|-------------------------------------|
-| enter-vr     | User has entered VR and fullscreen. |
-| exit-vr      | User has exited VR and fullscreen.  |
-| loaded       | All nodes have loaded.              |
+| enter-vr     | User has entered VR and headset starts presenting content. |
```
Comment (ngokevin) [inline docs/core/scene.md]: started

File: src/utils/urlOptions.js [hunk 29]
```diff
@@ -0,0 +1,20 @@
+module.exports = (function () {
```
Comment (ngokevin) [inline src/utils/urlOptions.js]: I created this as a small lib that's a bit more documented and tested https://github.com/ngokevin/urlgray/blob/master/index.js

If we don't want that, `queryParams` might be more accurate name than `urlOptions`

**Comment by dmarcos** [comment]
This is ready to merge.

File: src/style/aframe.css [hunk 30]
```diff
@@ -66,13 +71,13 @@ a-scene audio {
   display: flex;
   display: -webkit-flex;
   font-family: sans-serif, monospace;
-  font-size: 13px;
+  font-size: 13px;}
```
Comment (ngokevin) [inline src/style/aframe.css]: oops brace?

File: src/core/scene/postMessageAPI.js [hunk 31]
```diff
@@ -0,0 +1,30 @@
+var isIframed = require('../../utils/').isIframed;
```
Comment (ngokevin) [inline src/core/scene/postMessageAPI.js]: nit: could call everything `postMessage` instead of `postMessageAPI`.

**Pull request merged**